middle_end/flambda2/simplify/simplify_let_cont_expr.ml

```diff
@@ -1290,58 +1593,36 @@ let simplify_let_cont0 ~simplify_expr dacc (data : simplify_let_cont_data)
+  (* About scopes: supposing we are at scope 'n' before the let-cont, we will:
+
+     - use scope 'n + 2' to inspect the body
+
+     - use scope 'n + 1' to perform the join of continuation uses
+
+     - bump the scope of the env after join to get an env at scope 'n + 2'
+
+     - recursively for lifted continuations, whose join will take place at level
+     'n + 1' (which is the level for denv_for_join) *)
+  let denv_for_join = DE.increment_continuation_scope denv_before_body in
+  let denv_for_body = DE.increment_continuation_scope denv_for_join in
+  let denv_for_body =
+    DE.add_continuation_defined_in_current_continuation denv_for_body
   in
+  let dacc = DA.with_denv dacc denv_for_body in
   let body = data.body in
   let data : after_downwards_traversal_of_body_data =
-    { denv_before_body; prior_lifted_constants; handlers = data.handlers }
+    { denv_for_join; prior_lifted_constants; handlers = data.handlers }
   in
   simplify_expr dacc body
     ~down_to_up:
       (after_downwards_traversal_of_body ~simplify_expr data ~down_to_up)
 
-let simplify_let_cont ~simplify_expr dacc (let_cont : Let_cont.t) ~down_to_up =
+let simplify_let_cont ~simplify_expr dacc let_cont ~down_to_up =
   (* This is the entry point to simplify a let cont expression. The only thing
      it does is to match all handlers to break the name abstraction, and then
      call [simplify_let_cont_stage1]. *)
-  let data : simplify_let_cont_data =
-    match let_cont with
-    | Non_recursive { handler; _ } ->
-      let cont, body =
-        Non_recursive_let_cont_handler.pattern_match handler
-          ~f:(fun cont ~body -> cont, body)
-      in
-      let cont_handler = Non_recursive_let_cont_handler.handler handler in
-      let is_exn_handler = CH.is_exn_handler cont_handler in
-      let is_cold = CH.is_cold cont_handler in
-      let params, handler =
-        CH.pattern_match cont_handler ~f:(fun params ~handler ->
-            params, handler)
-      in
-      { body;
-        handlers =
-          Non_recursive { cont; params; handler; is_exn_handler; is_cold }
-      }
-    | Recursive handlers ->
-      let invariant_params, body, rec_handlers =
-        Recursive_let_cont_handlers.pattern_match handlers
-          ~f:(fun ~invariant_params ~body rec_handlers ->
-            invariant_params, body, rec_handlers)
-      in
-      assert (not (Continuation_handlers.contains_exn_handler rec_handlers));
-      let handlers = Continuation_handlers.to_map rec_handlers in
-      let continuation_handlers =
-        Continuation.Map.map
-          (fun handler ->
-            let is_cold = CH.is_cold handler in
-            CH.pattern_match handler ~f:(fun params ~handler ->
-                { params; handler; is_cold }))
-          handlers
-      in
-      { body; handlers = Recursive { invariant_params; continuation_handlers } }
-  in
-  simplify_let_cont0 ~simplify_expr dacc data ~down_to_up
+  let body, handlers = split_let_cont let_cont in
```

Comment:
Just above the line: the comment is obsolete. It seems it was before this PR.

Reply:
`split_let_cont` is not large and the name is not very informative. This code would probably be better with it inlined

Reply:
Maybe not inline, rather find a better nam.

Reply:
I manually inlined  `split_let_cont`. I think that at one point, `split_let_cont` was used at multiple points in the file, but that has now stopped being the case, so indeed there is no reason to keep it as a separate function.